### PR TITLE
Add joins on the orders list query builder conditionally

### LINKED ISSUE #14781: Admin order list heavy query
**Sylius version affected**: 1.12.1

**Description**  
When sylius admin orders page fetch orders, it joins unnecessary tables which makes query larger, heavier and longer.

**Steps to reproduce**  
Visit `/admin/orders/` page and check query created via doctrine. 
1. For counting orders it selects all unnecessary order data + joins order items, products and product variants. This query requires only order id count and order state. 
2. Displaying paginated order list. Same problem, fetches too much data + joins products and product variants. 

**Possible Solution**  
Changes must be made here:
`/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php`

From:
```
    public function createSearchListQueryBuilder(): QueryBuilder
    {
        return $this->createListQueryBuilder()
            ->leftJoin('o.items', 'item')
            ->leftJoin('item.variant', 'variant')
            ->leftJoin('variant.product', 'product')
        ;
    }
``` 

To:
```
    public function createSearchListQueryBuilder(?array $criteria = []): QueryBuilder
    {
        $builder = $this->createListQueryBuilder();

        $hasProductCriteria = $criteria['product'] ?? false;
        $hasVariantCriteria = $criteria['variant'] ?? false;

        if ($hasProductCriteria || $hasVariantCriteria) {
            $builder
                ->leftJoin('o.items', 'item')
                ->leftJoin('item.variant', 'variant')
                ->leftJoin('variant.product', 'product');
        }

        return $builder;
    }
```

### PR DESCRIPTION
**Pull request by jakubtobiasz**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13
| Bug fix?        | no/
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | closes #14781
| License         | MIT

### CODE REVIEW

**Comment by vvasiloi** [comment]
WDYT about adding a new  `createCriteriaSearchListQueryBuilder` method and deprecating the current one instead of changing the signature?

**Comment by jakubtobiasz** [comment]
@vvasiloi Seems like a safer way 👍🏼. I'll change it tomorrow.

File: UPGRADE-1.13.md [hunk 1]
```diff
@@ -114,3 +114,7 @@
 
    All internal usages of service `sylius.product_variant_resolver.default` have been switched to `Sylius\Component\Product\Resolver\ProductVariantResolverInterface`, if you have been using the
    `sylius.product_variant_resolver.default` service apply this change accordingly.
+
+1. Due to optimizations of the Order's grid the `\Sylius\Component\Core\Repository\OrderRepositoryInterface::createSearchListQueryBuilder` method bas been deprecated and replaced by `\Sylius\Component\Core\Repository\OrderRepositoryInterface::createCriteriaAwareSearchListQueryBuilder`.
+   Also `\Sylius\Component\Core\Repository\OrderRepositoryInterface::createByCustomerIdQueryBuilder` has been deprecated and replaced by `\Sylius\Component\Core\Repository\OrderRepositoryInterface::createByCustomerIdCriteriaAwareQueryBuilder` due to same reason. Both changes affected
+   `sylius_admin_order` and `sylius_admin_customer_order` grids configuration.
```
Comment (NoResponseMate) [inline UPGRADE-1.13.md]:
```suggestion
1. Due to optimizations of the Order's grid the `Sylius\Component\Core\Repository\OrderRepositoryInterface::createSearchListQueryBuilder` method bas been deprecated and replaced by `Sylius\Component\Core\Repository\OrderRepositoryInterface::createCriteriaAwareSearchListQueryBuilder`.
   Also `Sylius\Component\Core\Repository\OrderRepositoryInterface::createByCustomerIdQueryBuilder` has been deprecated and replaced by `Sylius\Component\Core\Repository\OrderRepositoryInterface::createByCustomerIdCriteriaAwareQueryBuilder` for the same reason. Both changes affect
   `sylius_admin_order` and `sylius_admin_customer_order` grids configuration.
```

**Review by NoResponseMate — DISMISSED** [review]

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Jacob! :tada:

File: src/Sylius/Bundle/AdminBundle/Resources/config/grids/customer_order.yml [hunk 2]
```diff
@@ -5,7 +5,7 @@ sylius_grid:
             driver:
                 options:
                     repository:
-                        method: createByCustomerIdQueryBuilder
+                        method: createByCustomerIdCriteriaAwareQueryBuilder
                         arguments:
                             customerId: $id
```
Comment (mamazu) [inline src/Sylius/Bundle/AdminBundle/Resources/config/grids/customer_order.yml:10]: Hey, I have a question about this code. This shouldn't work. The fist argument of the function is the criteria but in here you specify the customer id first. See:

[Grid bundle driver.](https://github.com/Sylius/SyliusGridBundle/blob/61970651ba4cf6547cb6b0d3220d7c9891f5e70d/src/Bundle/Doctrine/ORM/Driver.php#L53)

**Pull request merged**